### PR TITLE
The encoder and decoder for engineio v4 polling.

### PR DESCRIPTION
**Pull request by googollee**

### CODE REVIEW

File: engineio/transport/polling/polling_test.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package polling
+
+import (
+	"github.com/googollee/go-socket.io/engineio/frame"
+)
+
+type packet struct {
```
Comment (mrfoe7) [inline engineio/transport/polling/polling_test.go:7]: 👍

Comment (mrfoe7) [reply engineio/transport/polling/polling_test.go:7]: maybe we need public access in the future

**Comment by mrfoe7** [comment]
I'll review this code on this weekends

File: engineio/transport/polling/encoder.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	}
+
+	<-e.hasFramesChan
+	select {
+	case e.hasFramesChan <- struct{}{}:
+	default: // if it already has frames, the chan is full and continues.
+	}
+}
+
+type frameWriter struct {
```
Comment (mrfoe7) [inline engineio/transport/polling/encoder.go:92]: we need separate frameWriter and encoder by two different file in one package

File: engineio/transport/polling/decoder.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+		if err := d.lastFrame.reader.PushBack(1); err != nil {
+			return 0, nil, err
+		}
+	}
+
+	return frame.Text, d.lastFrame, nil
+}
+
+// frameReader is a reader to read one frame.
+type frameReader struct {
```
Comment (mrfoe7) [inline engineio/transport/polling/decoder.go:68]: i suggest, we need separate `frameReader` and encoder by two different file in one package

File: engineio/transport/polling/encoder.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+)
+
+// encoder encodes frames to the writer.
+type encoder struct {
+	pingTimeout       time.Duration
+	lastPing          time.Time
+	writer            *bufWriter
+	hasFramesChan     chan struct{}
+	closed            chan struct{}
+	hasNonClosedFrame int32
```
Comment (mrfoe7) [inline engineio/transport/polling/encoder.go:19]: `hasNonClosedFrame` is it very difficult read to this. i suggest rename this variable to `isClosedFrame`. What do you mean?

File: engineio/transport/polling/encoder.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+type frameWriter struct {
+	writer            *bufWriter
+	base64            io.WriteCloser
+	hasFramesChan     chan struct{}
+	hasNonClosedFrame *int32
+}
+
+func (w *frameWriter) Write(b []byte) (int, error) {
```
Comment (mrfoe7) [inline engineio/transport/polling/encoder.go:99]: i suggest change `b` to `data` into `Write` interfaces

File: engineio/transport/polling/encoder.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+	lastPing          time.Time
+	writer            *bufWriter
+	hasFramesChan     chan struct{}
+	closed            chan struct{}
+	hasNonClosedFrame int32
+}
+
+func newEncoder(pingTimouet time.Duration, closed chan struct{}, buf []byte) *encoder {
+	return &encoder{
+		pingTimeout:   pingTimouet,
```
Comment (mrfoe7) [inline engineio/transport/polling/encoder.go:24]: typo - `pingTimouet`

File: engineio/transport/polling/decoder.go [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+	reader   *bufReader
+	finished bool
+}
+
+// Read reads data of the frame to buffer b.
+func (r *frameReader) Read(b []byte) (int, error) {
```
Comment (mrfoe7) [inline engineio/transport/polling/decoder.go:74]: ?

File: engineio/transport/polling/decoder.go [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+	if n == 0 && err == nil {
+		return 0, io.EOF
+	}
+
+	return n, err
+}
+
+// ReadByte reads a byte in the frame.
+func (r *frameReader) ReadByte() (byte, error) {
```
Comment (mrfoe7) [inline engineio/transport/polling/decoder.go:98]: why are you need this method?

File: engineio/transport/polling/decoder.go [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+		return 0, io.EOF
+	}
+
+	return ret, nil
+}
+
+// Discard discards all data in the frame.
+func (r *frameReader) Discard() error {
+	var buf [1024]byte
+	for {
```
Comment (mrfoe7) [inline engineio/transport/polling/decoder.go:117]: we should change this part to
var err error
```go
for err != nil {
_, err = r.Read(buf[:])
}
return err
```

File: engineio/transport/polling/encoder_test.go [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+	}
+
+	n, err := wr.Write([]byte("1234"))
+	if err != nil || n != 4 {
+		t.Fatalf("wr.Write() want: (4, nil), got: (%d, %s)", n, err)
+	}
+
+	wantWait := time.Second / 10
+	var wg sync.WaitGroup
+	defer wg.Wait()
```
Comment (mrfoe7) [inline engineio/transport/polling/encoder_test.go:264]: ?

File: engineio/transport/polling/decoder.go [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+	"github.com/googollee/go-socket.io/engineio/frame"
+)
+
+// decoder decodes frames from the reader.
+type decoder struct {
+	reader    *bufReader
+	lastFrame *frameReader
+}
+
+func newDecoder(buf []byte, r io.Reader) *decoder {
```
Comment (mrfoe7) [inline engineio/transport/polling/decoder.go:17]: i suggest change `buf` to `data` into `newDecoder ` interface

File: engineio/transport/polling/decoder.go [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+	return &decoder{
+		reader: newBufReader(buf, r),
+	}
+}
+
+// NextFrame returns a reader to read the next frame.
+func (d *decoder) NextFrame() (frame.Type, io.Reader, error) {
```
Comment (mrfoe7) [inline engineio/transport/polling/decoder.go:24]: i suggest return the next type from method:

```go
type FrameReader struct {
	ft frame.Type
	io.Reader
}
```

what do u think about it?

**Comment by mrfoe7** [comment]
pls, add package `github.com/google/go-cmp v0.5.6` in go.mod
Снимок экрана 2021-07-12 в 14 03 12

File: engineio/transport/polling/decoder.go [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+		// Start from the 2nd frame, need to ignore the separator.
+		next, err = d.reader.ReadByte()
+		if err != nil && err != io.EOF {
+			return 0, nil, err
+		}
+	}
+
+	d.lastFrame = &frameReader{
+		reader:   d.reader,
+		finished: false,
```
Comment (mrfoe7) [inline engineio/transport/polling/decoder.go:50]: not important comment. 

`finished: false` - it is default value. it does not necessary

File: engineio/transport/polling/decoder.go [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+		}
+	}
+
+	return frame.Text, d.lastFrame, nil
+}
+
+// frameReader is a reader to read one frame.
+type frameReader struct {
+	reader   *bufReader
+	finished bool
```
Comment (mrfoe7) [inline engineio/transport/polling/decoder.go:70]: why are you need this flag?

File: engineio/transport/polling/decoder.go [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+		if b[i] == separator {
+			r.finished = true
+			if err := r.reader.PushBack(n - i); err != nil {
+				return n, fmt.Errorf("decode package error:(it should not happen) %w", err)
+			}
+			n = i
+		}
+	}
+
+	if err == io.EOF {
```
Comment (mrfoe7) [inline engineio/transport/polling/decoder.go:86]: i think, this errors checking from line 86-92 need move after 75 line

```go
n, err := r.reader.Read(b)
if (n == 0 && err == nil) || err == io.EOF {
    return 0, io.EOF
}
if err != nil {
    return 0, err
}
```

File: engineio/transport/polling/encoder.go [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+type encoder struct {
+	pingTimeout       time.Duration
+	lastPing          time.Time
+	writer            *bufWriter
+	hasFramesChan     chan struct{}
+	closed            chan struct{}
+	hasNonClosedFrame int32
+}
+
+func newEncoder(pingTimouet time.Duration, closed chan struct{}, buf []byte) *encoder {
```
Comment (mrfoe7) [inline engineio/transport/polling/encoder.go:22]: why are you need trow `closed` channel and `pingTimeout` in function interface?

i think `pingTimeout` is optional value

File: engineio/transport/polling/encoder.go [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+
+	<-e.hasFramesChan
+	select {
+	case e.hasFramesChan <- struct{}{}:
+	default: // if it already has frames, the chan is full and continues.
+	}
+}
+
+type frameWriter struct {
+	writer            *bufWriter
```
Comment (mrfoe7) [inline engineio/transport/polling/encoder.go:93]: we should join `writer` and `base64` filed by one interface `io.WriteCloser`. it is simple

File: engineio/transport/polling/decoder.go (hunk 3 above)
Comment (googollee) [reply engineio/transport/polling/decoder.go:68]: I did before, but as `decoder` and `frameReader` are so closed to each other, I need to switch between two files often. So I put them together.

File: engineio/transport/polling/encoder.go (hunk 4 above)
Comment (googollee) [reply engineio/transport/polling/encoder.go:19]: `hasNonClosedFrame` This means the caller requested a frame to write but didn't close it yet. It's the opposite meaning of `isClosedFrame`.

File: engineio/transport/polling/decoder.go (hunk 8 above)
Comment (googollee) [reply engineio/transport/polling/decoder.go:98]: `ReadByte()` is useful when handling frames. For binary frames, they have a prefix byte 'b'. Also in socket.io layer,a text frame has a lot of metadata which is suite for 'ReadByte()'.

File: engineio/transport/polling/decoder.go (hunk 15 above)
Comment (googollee) [reply engineio/transport/polling/decoder.go:86]: Not really. `io.Reader.Read()` doesn't promise `n == 0` when`err != nil`. So to follow this , need to check b with n before checking err.

See https://pkg.go.dev/io#Reader.

File: engineio/transport/polling/encoder_test.go (hunk 10 above)
Comment (googollee) [reply engineio/transport/polling/encoder_test.go:264]: Please say something. It's hard to guess with a question mark.

File: engineio/transport/polling/decoder.go (hunk 11 above)
Comment (googollee) [reply engineio/transport/polling/decoder.go:17]: For this parameter, it's not used as data bytes, but a pre-allocated buffer for the decoder.

File: engineio/transport/polling/encoder.go (hunk 2 above)
Comment (googollee) [reply engineio/transport/polling/encoder.go:92]: See the comment of `decoder` and `frameReader`.

File: engineio/transport/polling/encoder.go (hunk 16 above)
Comment (googollee) [reply engineio/transport/polling/encoder.go:22]: No. Encoder handles timeout in `WriteFramesTo()`. The transport will block on `WriteFramesTo()` to write something out. If it doesn't have data to write, then returns timeout error to process timeout.

File: engineio/transport/polling/decoder.go (hunk 14 above)
Comment (googollee) [reply engineio/transport/polling/decoder.go:70]: Yes, we could remove it.

File: engineio/transport/polling/encoder.go (hunk 17 above)
Comment (googollee) [reply engineio/transport/polling/encoder.go:93]: I don't understand this suggestion. Could you give an example?

File: engineio/transport/polling/encoder.go (hunk 5 above)
Comment (googollee) [reply engineio/transport/polling/encoder.go]: Done.

File: engineio/transport/polling/encoder.go (hunk 6 above)
Comment (googollee) [reply engineio/transport/polling/encoder.go]: Done.

File: engineio/transport/polling/decoder.go (hunk 9 above)
Comment (googollee) [reply engineio/transport/polling/decoder.go]: Done.

File: engineio/transport/polling/decoder.go (hunk 12 above)
Comment (googollee) [reply engineio/transport/polling/decoder.go]: Done.

File: engineio/transport/polling/decoder.go (hunk 13 above)
Comment (googollee) [reply engineio/transport/polling/decoder.go]: Done.

**Comment by googollee** [comment]
Fixed `go.mod`.

File: engineio/transport/polling/encoder.go (hunk 4 above)
Comment (mrfoe7) [reply engineio/transport/polling/encoder.go:19]: >> It's the opposite meaning of isClosedFrame.

Yeap, sure. you'll change current logic to opposite

File: engineio/transport/polling/encoder.go (hunk 17 above)
Comment (mrfoe7) [reply engineio/transport/polling/encoder.go:93]: My suggest is:
```go
type frameWriter struct {
writer io.WriteCloser
hasFramesChan chan struct {}
isClosedFrame *int32
}

i added changes to next commit

File: engineio/transport/polling/decoder.go (hunk 15 above)
Comment (mrfoe7) [reply engineio/transport/polling/decoder.go:89]: oh, thanks

File: engineio/transport/polling/encoder.go (hunk 17 above)
Comment (mrfoe7) [reply engineio/transport/polling/encoder.go]: oh, we need second writers by `frameWriter`. i rollbacked my commit

File: engineio/transport/polling/encoder.go (hunk 16 above)
Comment (mrfoe7) [reply engineio/transport/polling/encoder.go]: i suggest change signature to:
```go
func newEncoder(buf []byte,  closed chan struct{}, pingTimeout time.Duration) *encoder {
```
i mean that use config/options type into encode struct with default `pintTimeout`, but now we use one optional value. at now, it is doesn't necessary

**Comment by mrfoe7** [comment]
LGTM

**Review by mrfoe7 — APPROVED** [review]

File: engineio/transport/polling/encoder.go (hunk 4 above)
Comment (googollee) [reply engineio/transport/polling/encoder.go:19]: No. Doing that makes the whole logic complex. When creating a new encoder, `hasNonClosedFrame` is 0 by default, which is natural. When someone calls `Encoder.NextFrame()`, `hasNonClosedFrame` increases.

I know `hasClosedFrame` sounds better, but it's hard to use 0 as default value.

**Pull request merged**